File: lib/domain/world/world.dart
import 'coords.dart';
import 'settlement.dart';
import 'tile.dart';
import 'world_gen.dart';

/// Um bloco de tiles já resolvidos, pronto para o renderizador consumir.
class WorldChunk {
  WorldChunk({required this.coord, required List<WorldTile> tiles})
      : _tiles = tiles,
        assert(
          tiles.length == WorldMetrics.chunkSize * WorldMetrics.chunkSize,
          'chunk deve ter chunkSize^2 tiles',
        );

  final ChunkCoord coord;
  final List<WorldTile> _tiles;

  WorldTile tileAt(int localX, int localY) =>
      _tiles[localY * WorldMetrics.chunkSize + localX];

  /// Itera os tiles em ordem de profundidade isométrica (do fundo para a
  /// frente), que é a ordem em que o renderizador precisa desenhar.
  Iterable<(TileCoord, WorldTile)> get tilesInDepthOrder sync* {
    final origin = coord.origin;
    const size = WorldMetrics.chunkSize;
    for (var sum = 0; sum <= (size - 1) * 2; sum++) {
      for (var ly = 0; ly < size; ly++) {
        final lx = sum - ly;
        if (lx < 0 || lx >= size) continue;
        yield (
          TileCoord(origin.x + lx, origin.y + ly),
          _tiles[ly * size + lx],
        );
      }
    }
  }
}

/// Fachada sobre o gerador: resolve tiles sob demanda e mantém um cache LRU de
/// chunks para que o pan da câmera não regenere ruído a cada frame.
class World {
  World({required this.generator, required this.layout});

  /// Cria o mundo de uma campanha nova a partir da seed.
  factory World.fromSeed(int seed) {
    final generator = WorldGenerator(seed: seed);
    return World(generator: generator, layout: generator.generateLayout());
  }

  /// Recria o mundo de uma campanha salva. O terreno é regenerado da seed; só
  /// o layout vem do save.
  factory World.restore({required int seed, required WorldLayout layout}) =>
      World(generator: WorldGenerator(seed: seed), layout: layout);

  final WorldGenerator generator;
  final WorldLayout layout;

  /// 256 chunks ~= 65k tiles. Cobre várias telas de pan em celular sem
  /// pressionar a memória.
  static const int _maxCachedChunks = 256;

  final Map<ChunkCoord, WorldChunk> _cache = {};
  final List<ChunkCoord> _lru = [];

  int get seed => generator.seed;

  WorldTile tileAt(int x, int y) {
    final coord = TileCoord(x, y);
    final chunk = chunkAt(coord.chunk);
    return chunk.tileAt(coord.localX, coord.localY);
  }

  WorldChunk chunkAt(ChunkCoord coord) {
    final cached = _cache[coord];
    if (cached != null) {
      _touch(coord);
      return cached;
    }

    final chunk = _generateChunk(coord);
    _cache[coord] = chunk;
    _lru.add(coord);
    _evictIfNeeded();
    return chunk;
  }

  void _touch(ChunkCoord coord) {
    _lru.remove(coord);
    _lru.add(coord);
  }

  void _evictIfNeeded() {
    while (_lru.length > _maxCachedChunks) {
      _cache.remove(_lru.removeAt(0));
    }
  }

  WorldChunk _generateChunk(ChunkCoord coord) {
    const size = WorldMetrics.chunkSize;
    final origin = coord.origin;
    final tiles = <WorldTile>[];
    for (var ly = 0; ly < size; ly++) {
      for (var lx = 0; lx < size; lx++) {
        tiles.add(generator.tileAt(origin.x + lx, origin.y + ly, layout));
      }
    }
    return WorldChunk(coord: coord, tiles: tiles);
  }

  /// Descarta o cache. Útil ao trocar de campanha.
  void clearCache() {
    _cache.clear();
    _lru.clear();
  }

  int get cachedChunkCount => _cache.length;

  // ---------------------------------------------------------------------------
  // Consultas de jogo
  // ---------------------------------------------------------------------------

  Settlement? settlementAt(TileCoord tile) => layout.settlementAt(tile);

  /// Assentamento mais próximo de um tile, em linha reta.
  Settlement nearestSettlement(TileCoord tile) {
    var best = layout.settlements.first;
    var bestDistance = double.infinity;
    for (final s in layout.settlements) {
      final d = tile.euclideanTo(s.center);
      if (d < bestDistance) {
        bestDistance = d;
        best = s;
      }
    }
    return best;
  }

  /// Recursos extraíveis num raio, agregados por item. Alimenta a tela de
  /// "o que dá para trabalhar por aqui".
  Map<String, double> surveyResources(TileCoord center, {int radius = 12}) {
    final totals = <String, double>{};
    for (var dy = -radius; dy <= radius; dy++) {
      for (var dx = -radius; dx <= radius; dx++) {
        final tile = tileAt(center.x + dx, center.y + dy);
        if (!tile.hasResource) continue;
        final key = tile.resource!.name;
        totals[key] = (totals[key] ?? 0) + tile.resourceRichness;
      }
    }
    return totals;
  }
}

File: web3d/src/world/world.ts
import { ChunkCoord, CHUNK_SIZE, TileCoord } from './coords';
import { WorldLayout, ensureRoadPaths, generateLayout, tileAt } from './layout';
import type { Settlement } from './settlement';
import { hasResource, type WorldTile } from './tile';
import { WorldGenerator } from './worldGen';

/**
 * Fachada sobre o gerador: resolve tiles sob demanda e mantém um cache de
 * chunks para o pan da câmera não regenerar ruído a cada quadro.
 */

/** Um bloco de tiles já resolvidos, pronto para o renderizador consumir. */
export class WorldChunk {
  constructor(
    readonly coord: ChunkCoord,
    private readonly tiles: readonly WorldTile[],
  ) {
    if (tiles.length !== CHUNK_SIZE * CHUNK_SIZE) {
      throw new Error(`chunk deve ter ${CHUNK_SIZE}² tiles`);
    }
  }

  tileAt(localX: number, localY: number): WorldTile {
    return this.tiles[localY * CHUNK_SIZE + localX]!;
  }

  /**
   * Itera em ordem de profundidade isométrica, do fundo para a frente.
   *
   * O renderizador 3D não precisa disso — ele tem z-buffer —, mas o cliente
   * isométrico precisava, e a ordem continua sendo a definição de "fundo" para
   * qualquer coisa que empilhe sprites. Sai de graça: é a mesma varredura por
   * diagonal.
   */
  *tilesInDepthOrder(): Generator<[TileCoord, WorldTile]> {
    const origin = this.coord.origin;
    for (let sum = 0; sum <= (CHUNK_SIZE - 1) * 2; sum++) {
      for (let ly = 0; ly < CHUNK_SIZE; ly++) {
        const lx = sum - ly;
        if (lx < 0 || lx >= CHUNK_SIZE) continue;
        yield [
          new TileCoord(origin.x + lx, origin.y + ly),
          this.tiles[ly * CHUNK_SIZE + lx]!,
        ];
      }
    }
  }
}

/** 256 chunks ≈ 65 mil tiles: várias telas de pan sem pressionar a memória. */
const MAX_CACHED_CHUNKS = 256;

export class World {
  private readonly cache = new Map<string, WorldChunk>();
  /**
   * Ordem de uso, do mais antigo para o mais recente.
   *
   * Lista e não `Map` com ordem de inserção porque "tocar" um item exige
   * removê-lo do meio, e um `Map` não sabe fazer isso sem `delete` + `set` —
   * que é exatamente o mesmo custo com mais chance de erro.
   */
  private readonly lru: string[] = [];

  constructor(
    readonly generator: WorldGenerator,
    readonly layout: WorldLayout,
  ) {}

  /** Cria o mundo de uma campanha nova a partir da seed. */
  static fromSeed(seed: number): World {
    const generator = new WorldGenerator(seed);
    return new World(generator, generateLayout(generator));
  }

  /**
   * Recria o mundo de uma campanha salva.
   *
   * O terreno é regenerado da seed; só o layout vem do save. Guardar tiles seria
   * salvar o que já se sabe calcular, e o arquivo cresceria sem limite.
   */
  static restore(seed: number, layout: WorldLayout): World {
    const generator = new WorldGenerator(seed);
    // O save não guarda o traçado das estradas — ele é função pura da seed, e
    // guardá-lo custava 4 MB por mundo. Reconstruir aqui é o que faz o mapa
    // continuar desenhando as rotas depois de recarregar.
    return new World(generator, ensureRoadPaths(layout, generator));
  }

  get seed(): number {
    return this.generator.seed;
  }

  tileAt(x: number, y: number): WorldTile {
    const coord = new TileCoord(x, y);
    return this.chunkAt(coord.chunk).tileAt(coord.localX, coord.localY);
  }

  chunkAt(coord: ChunkCoord): WorldChunk {
    const chave = coord.key;
    const cached = this.cache.get(chave);
    if (cached) {
      this.touch(chave);
      return cached;
    }

    const chunk = this.generateChunk(coord);
    this.cache.set(chave, chunk);
    this.lru.push(chave);
    this.evictIfNeeded();
    return chunk;
  }

  private touch(chave: string): void {
    const i = this.lru.indexOf(chave);
    if (i >= 0) this.lru.splice(i, 1);
    this.lru.push(chave);
  }

  private evictIfNeeded(): void {
    while (this.lru.length > MAX_CACHED_CHUNKS) {
      this.cache.delete(this.lru.shift()!);
    }
  }

  private generateChunk(coord: ChunkCoord): WorldChunk {
    const origin = coord.origin;
    const tiles: WorldTile[] = [];
    for (let ly = 0; ly < CHUNK_SIZE; ly++) {
      for (let lx = 0; lx < CHUNK_SIZE; lx++) {
        tiles.push(tileAt(this.generator, origin.x + lx, origin.y + ly, this.layout));
      }
    }
    return new WorldChunk(coord, tiles);
  }

  /** Descarta o cache. Útil ao trocar de campanha. */
  clearCache(): void {
    this.cache.clear();
    this.lru.length = 0;
  }

  get cachedChunkCount(): number {
    return this.cache.size;
  }

  // ------------------------------------------------------ consultas de jogo

  settlementAt(tile: TileCoord): Settlement | null {
    return this.layout.settlementAt(tile);
  }

  /** Assentamento mais próximo, em linha reta. */
  nearestSettlement(tile: TileCoord): Settlement {
    let melhor = this.layout.settlements[0]!;
    let menor = Infinity;
    for (const s of this.layout.settlements) {
      const d = tile.euclideanTo(s.center);
      if (d < menor) {
        menor = d;
        melhor = s;
      }
    }
    return melhor;
  }

  /**
   * Recursos extraíveis num raio, agregados por item.
   *
   * Alimenta a tela de "o que dá para trabalhar por aqui". O raio padrão de 12
   * cobre um dia de caminhada — perguntar por um raio maior descreve um lugar
   * onde o jogador não vai conseguir chegar e voltar no mesmo reset.
   */
  surveyResources(center: TileCoord, radius = 12): Record<string, number> {
    const totais: Record<string, number> = {};
    for (let dy = -radius; dy <= radius; dy++) {
      for (let dx = -radius; dx <= radius; dx++) {
        const tile = this.tileAt(center.x + dx, center.y + dy);
        if (!hasResource(tile)) continue;
        const chave = tile.resource!;
        totais[chave] = (totais[chave] ?? 0) + tile.resourceRichness;
      }
    }
    return totais;
  }
}
